stock_picking_auto_create_lot/models/stock_move_line.py

```diff
@@ -1,42 +1,18 @@
 # Copyright 2020 ACSONE SA/NV
+# Copyright 2024 Tecnativa - Sergio Teruel
 # License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
 from odoo import models
-from odoo.fields import first
 
 
 class StockMoveLine(models.Model):
 
     _inherit = "stock.move.line"
 
-    def _prepare_auto_lot_values(self):
-        """
-        Prepare multi valued lots per line to use multi creation.
-        """
-        self.ensure_one()
-        return {"product_id": self.product_id.id, "company_id": self.company_id.id}
-
     def set_lot_auto(self):
         """
-        Create lots using create_multi to avoid too much queries
-        As move lines were created by product or by tracked 'serial'
-        products, we apply the lot with both different approaches.
+        Set lot_name field in stock move line to let Odoo the creation of lots
         """
-        values = []
-        production_lot_obj = self.env["stock.production.lot"]
-        lots_by_product = dict()
-        for line in self:
-            values.append(line._prepare_auto_lot_values())
-        lots = production_lot_obj.create(values)
-        for lot in lots:
-            if lot.product_id.id not in lots_by_product:
-                lots_by_product[lot.product_id.id] = lot
-            else:
-                lots_by_product[lot.product_id.id] += lot
-        for line in self:
-            lot = first(lots_by_product[line.product_id.id])
-            line.lot_id = lot
-            # If product has tracking we have to remove the lot to assign the next
-            # lot created. If not do it, different stock move lines will have the same
-            # lot and we will have orphans lots.
-            if lot.product_id.tracking != "none":
-                lots_by_product[line.product_id.id] -= lot
+        for line in self.filtered(
+            lambda sml: not sml.lot_name and sml.tracking != "none"
+        ):
+            line.lot_name = self.env["ir.sequence"].next_by_code("stock.lot.serial")
```

Comment:
I think `filtered()` here is redundant as `_set_auto_lot()` in stock.picking has already done this filtering.  I guess we could simply remove this method and complete `lot_name` assignment in the method in the picking.

It would be good to put `self.env["ir.sequence"].next_by_code("stock.lot.serial")` part into a method for extensibility.  For example, a client of ours uses the `product_lot_sequence` module along with this module and they currently work nicely together (auto-created lots/serials follow the sequence defined in `product_lot_sequence`), but the suggested change (i.e. assign the `lot_name` instead of creating a lot) will "break" this design, although these two modules may have not been intended to work together in the first place.  Having a hook point will allow us to create a bridge module to restore the behavior as necessary.

We created https://github.com/OCA/stock-logistics-workflow/pull/1561 for 16.0 with a less radical fix for the addressed issue, by the way. But I like the approach suggested in this PR better as it probably lightens the maintenance burden for the future.